### DISCUSSION TITLE
Failure to build wheel due to missing file.

### DISCUSSION THREAD

**Discussion by NeonFlames** [pypa/setuptools#3748] [Q&A] [answered]
While changing from setup.py to pyproject.toml I encountered an error where files cannot be found while building the wheel.
https://github.com/Tuxemon/Tuxemon/actions/runs/3834153146/jobs/6526274169 shows the error during an editable install and the following error occurs when trying to build via python -m build: error: could not create 'build/bdist.linux-x86_64/wheel/tuxemon/cli/commands/__init__.py': No such file or directory

Have I missed a configuration option or is it something else?

**Comment by webknjaz** [comment]
Maybe include it through `MANIFEST.in` (in case you use `setuptools-scm`, this shouldn't be necessary).
FWIW `python -m build` without args builds an sdist and then a wheel out of that sdist. So if you forget to include files in sdist, there could be not enough sources to make a wheel out of it.
You could try building just an sdist locally (`python -m build --sdist`) and then inspecting what's actually inside (you can unpack the tarball via `tar xzvf dist/*.tar.gz`).
Also, maybe try converting to `setup.cfg` first instead of going straight to the `pyproject.toml`? (you'll still need a PEP 517 section in the latter — start with that)

**Comment by abravalheri** [comment, marked as answer]
Hi @NeonFlames I suggest the following changes in the code base:

```diff
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -37,11 +37,10 @@ tuxemon = "tuxemon.__main__:main"
 include-package-data = true

 [tool.setuptools.packages.find]
-where = [""]
+where = ["."]
 include = ["tuxemon*"]

 [tool.setuptools.dynamic]
-version = {attr = "setup.get_version"}
 dependencies = {file = "requirements.txt"}

 [tool.black]
diff --git a/setup.py b/setup.py
index a55ea254..fce63d92 100644
--- a/setup.py
+++ b/setup.py
@@ -8,4 +8,4 @@ def get_version():


 # Configure setuptools
-setup()
+setup(version=get_version())
```

Please note that `version = {attr = ...}` will get a value by reading an attribute from a module and this value will be used directly as version. The module should be *inside* of the package you are creating, and the value should ideally be compatible with `ast.literal_eval`.

You cannot provide a function, since setuptools will not execute the function before using it. See [note 4 on docs](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#attr).

I believe that is what was confusing the build process (I imagine it was getting somethings wrong and not creating the necessary folders before trying to copy the files).

Please also consider the points brought up by [webknjaz](https://github.com/webknjaz): if you have non-Python files in your repo, it is important to use either `setuptools-scm` (automatic way) or `MANIFEST.in` (manual way) to ensure all files are present in the distribution. See [section on docs](https://setuptools.pypa.io/en/latest/userguide/miscellaneous.html). Inspecting the `sdist`/`wheel` contents is a good way of checking that everything is there.

**Reply by NeonFlames** [reply]
You were right the version was the issue, we could probably manually set it since the only other place that has the version would be the README.

Also thank you both for the quick and helpful replies!

**Reply by NeonFlames** [reply]
The only thing I don't get is that the sdist worked just fine while the wheel build didn't.

**Reply by abravalheri** [reply]
I am not sure... Probably what is happening is that before "tar.gz"-ing the sdist `setuptools` write the files to a simply-named directory, while for the wheel the naming/nesting of that directory is more complex and probably involves the version number.

**Reply by NeonFlames** [reply]
It gets the version number correct as seen here https://pastebin.com/y5G888Ns for the sdist, but that makes sense for the wheel.

**Reply by NeonFlames** [reply]
Thank you both for your time, have a nice day!
